pkg/util/validation/validate.go

```diff
@@ -177,6 +189,42 @@ func ValidateExemplar(userID string, ls []mimirpb.LabelAdapter, e mimirpb.Exempl
 	return nil
 }
 
+// ValidateHistograms returns an err if the histogram is invalid.
+// The returned error may retain the provided series labels.
+// It uses the passed 'now' time to measure the relative time of the histogram.
+func ValidateHistogram(now model.Time, cfg SampleValidationConfig, userID string, ls []mimirpb.LabelAdapter, h mimirpb.Histogram) ValidationError {
+	unsafeMetricName, _ := extract.UnsafeMetricNameFromLabelAdapters(ls)
+
+	if model.Time(h.Timestamp) > now.Add(cfg.CreationGracePeriod(userID)) {
+		DiscardedSamples.WithLabelValues(reasonTooFarInFuture, userID).Inc()
+		return newSampleTimestampTooNewError("histogram", unsafeMetricName, h.Timestamp)
+	}
+
+	checkSpans := func(sign string, spans []*mimirpb.BucketSpan, buckets []int64) error {
+		var spanBuckets uint32
+		for n, span := range spans {
+			if n > 0 && span.Offset < 0 {
+				return newHistogramSpanNegativeOffsetError(sign, n+1, span.Offset, h.Timestamp, unsafeMetricName)
+			}
+			spanBuckets += span.Length
+		}
+		if l := uint32(len(buckets)); spanBuckets != l {
```

Comment:
Or declare spanBuckets as int
```suggestion
		if int(spanBuckets) != len(buckets) {
```